src/core.jl

```diff
@@ -121,51 +140,40 @@ eltype{T}(::Type{Signal{T}}) = T
 
 ##### Connections #####
 
-function add_action!(f, node, recipient)
-    a = Action(WeakRef(recipient), f)
+function add_action!(f, node)
+    a = Action(WeakRef(node), f)
     push!(node.actions, a)
     maybe_restart_queue()
     a
 end
 
-function remove_action!(f, node, recipient)
-    node.actions = filter(a -> a.f != f, node.actions)
+"""
+Removes `action` from `node.actions`
+"""
+function remove_action!(node, action)
+    filter!(a -> a != action, node.actions)
+    nothing
 end
 
-function close(n::Signal, warn_nonleaf=true)
-    finalize(n) # stop timer etc.
-    n.alive = false
-    if !isempty(n.actions)
-        any(map(isrequired, n.actions)) && warn_nonleaf &&
-            warn("closing a non-leaf node is not a good idea")
-        empty!(n.actions)
+"""
+Remove node from nodes - called as a finalizer on GC'd node
+"""
+function remove_node!(deadnode)
+    filter!(nodes) do node_ref
+        node_ref.value != deadnode
     end
+    nothing
+end
 
+function close(n::Signal, warn_nonleaf=true)
+    remove_node!(n)
     for p in n.parents
         delete!(p.preservers, n)
     end
+    finalize(n) # stop timer etc.
 end
 
-function send_value!(node::Signal, x, timestep)
-    # Dead node?
-    !node.alive && return
-
-    # Set the value and do actions
-    node.value = x
-    for action in node.actions
-        action.recipient.value != nothing && #nothing means downstream node has been gc'd
-            DataStructures.enqueue!(action_queue, (action.recipient.value, action))
-    end
-end
-send_value!(wr::WeakRef, x, timestep) = wr.value != nothing && send_value!(wr.value, x, timestep)
-
-do_action(a::Action, timestep) =
-    isrequired(a) && a.f(a.recipient.value, timestep)
-
-# If any actions have been gc'd, remove them
-cleanup_actions(node::Signal) =
-    node.actions = filter(isrequired, node.actions)
-
+send_value!(node::Signal, x) = (node.value = x)
```

Comment:
`set_value!` could be useful to isolate the state change. I imagine it would get inlined everywhere and not affect performance, but I'll check.